code/android-core-library/src/main/java/com/adobe/marketing/mobile/utils/RemoteDownloader.kt

```diff
@@ -0,0 +1,116 @@
+        private const val TAG = "RemoteDownloader"
+    }
+
+    /**
+     * Represents a component that provides metadata about file content.
+     */
+    interface MetadataProvider {
+        companion object MetadataKeys {
+            internal const val HTTP_HEADER_IF_MODIFIED_SINCE = "If-Modified-Since"
+            internal const val HTTP_HEADER_IF_RANGE = "If-Range"
+            internal const val HTTP_HEADER_RANGE = "Range"
+            internal const val HTTP_HEADER_LAST_MODIFIED = "Last-Modified"
+            internal const val ETAG = "ETag"
+        }
+
+        /**
+         * Retrieves metadata for the [file] provided for the purpose of
+         * conditionally fetching content from the remote url.
+         *
+         * @param file the [File] for which metadata is needed
+         * @return the metadata of the [file] if it is valid;
+         *         emptymap if metadata cannot be computed
+         */
+        fun getMetadata(file: File?): Map<String, String>
+    }
+
+    /**
+     * Allows easier mocking during tests.
+     */
+    private val downloadJobSupplier: (Networking, CacheFileService, url: String, downloadSubDirectory: String?, MetadataProvider) -> RemoteDownloadJob =
+        { networking, fileCacheService, url, downloadDirectory, metadataProvider ->
+            RemoteDownloadJob(
+                networking,
+                fileCacheService,
+                url,
+                downloadDirectory,
+                metadataProvider
+            )
+        }
+
+    /**
+     * Triggers content download from [url] to [downloadSubDirectory] by using the [metadataProvider]
+     * to facilitate any partial downloads.
+     * Note that currently, triggers to download are independent of successive calls.
+     * [download] when called with the same url without waiting for a result via [completionCallback]
+     * will cause ambiguous results.
+     *
+     * @param url the url from which the content must be downloaded from
+     * @param downloadSubDirectory optional directory within the default cache directory used by [cacheFileService]
+     *        into which the rules must be placed
+     * @param metadataProvider the metadata provider that will be used to fetch conditional headers when making a
+     *        download request
+     * @param completionCallback the callback to notify about downloaded result
+     */
+    fun download(
+        url: String,
+        downloadSubDirectory: String?,
+        metadataProvider: MetadataProvider,
+        completionCallback: (File?) -> Unit
+    ) {
+        if (!StringUtils.stringIsUrl(url)) {
+            MobileCore.log(
+                LoggingMode.WARNING,
+                TAG,
+                "Invalid URL: ($url). Contents cannot be downloaded."
+            )
+            completionCallback.invoke(null)
+            return
+        }
+
+        val remoteDownloadJob = downloadJobSupplier(
+            networkService,
+            cacheFileService,
```

Comment:
I think we may not cache the downloaded files inside this class. This is a public class exposed to all extensions, if we change any caching logic inside this class, it will affect other extensions. I think it should be better If we can let clients decide how/where to cache the downloaded files.
Also, we may need to add other parameters to the `download` function that shows the metadata of the previously downloaded files and how to handle 304 (not modified) response.

Reply:
Agree with eliminating the caching logic and deferring it to the caller of `download(...)`. 
However, based on discussion with @praveek , in the interest of merging Configuration extension, we planned to leave this as is, i.e 1:1 parity with existing production functionality and, then, create a new PR which eliminates caching as a fast follow to this PR. Does it sound good to you?